openedx_learning/core/itemstore/models_api.py

```diff
@@ -0,0 +1,21 @@
+from django.db import models
+
+from .models import ItemVersion
+
+
+class ItemVersionDataMixin(models.Model):
+    """
+    Minimal abstract model to let people attach data to ItemVersions.
+
+    The idea is that if you have a model that is associated with a specific
+    version of an item, the join is going to be 1:1 with an ItemVersion, and
```

Comment:
So, I'm reading that:
* each ItemVersion can only be associated with at most one of each type of "Data" model (Image, Asset, Video, etc.), but
* each ItemVersion may be associated with an arbitrary number of Content objects.

This confused me at first, because the "Data" models and the Content model occupy very similar roles in my head. But I think I'm beginning to understand it... is this a good example?
* Some ItemVersion represents a video.
  * It has multiple Content items associated with: one that points to the video file, one PNG thumbnail, and several others text files that are transcripts.
  * In the staticassets app, it has a single associated Video instance and Asset instance. Video metadata as well as information about the transcripts are part of (or hang off) of the Video model.

Reply:
Yeah, that was my intent. A couple of extra notes about it:

### Content survives after apps die.

At some point, an app becomes unsupported and disappears. At that point, all the Item/ItemVersion-related data it has dies as well. But Content is associated with the ItemVersion directly in `itemstore`, so that raw data will still export.

So I think this encourages a pattern where we store raw things in Content and then build smarter relational data structures out of them connected to ItemVersion. For instance, a ProblemBlock's raw OLX can live in Content while richer relational tables live in a new ProblemBlock app that give a richer, more queryable interface.

### ItemVersion mapping to just one row in a "Data" model is common, but not required.

I think that most of the time apps want to associate one piece of data with a given ItemVersion (e.g. a problem weight, schedule, etc.), but that's not a requirement. A rich data model of a ProblemBlock might want to create entries for each response, for instance.

Reply:
> ItemVersion mapping to just one row in a "Data" model is common, but not required.

How would that look?

For example, imagine we have an ItemVersion representing a problem block, and Response data models that we want to (indirectly) hang off of it. Would it look like:

```
ItemVersion  <--(1:1)-- ItemVersionProblemBlock --(N:N)--> Response
```
or:
```
ItemVersion  <--(1:N)-- ItemVersionResponse --(N:1)--> Response
```
or something else?

Reply:
I guess I thought it would be like:

```
ItemVersion <-- (N:1) -- ItemVersionProblemBlock -- (N:N) --> Response
```

That would still allow some central process to increment to the next version in a simple way without understanding what ItemVersionProblemBlock is. But I honestly haven't thought it through that much.